### ISSUE TITLE
Bug in `shouldContainSame` for Map

### ISSUE THREAD

**Issue by Murtaught** [MarkusAmshove/Kluent#170]
I'm quite sure there is a bug it `infix fun <K, V, M : Map<K, V>> M.shouldContainSame(expected: M): M`.

Judging by name it should raise exception if maps are not equal, but in fact this function does same thing as `shouldContainAll` but less efficiently, i.e. it checks that every key-value pair in `expected` is also in `this`, but not vice-versa.

Test:
 ```kotlin
mapOf("a" to 5, "b" to "Hello") shouldContainSame mapOf("a" to 5)
```
Passes, though is should fail.

I'm sorry if I didn't understand your intentions correctly. And, by the way, thanks for very useful library!

**Comment by Murtaught** [comment]
There is one more problem: whole reverse check is executed for every pair in `expected`.

I think fixed code should look like this:
```kotlin
infix fun <K, V, M : Map<K, V>> M.shouldContainSame(expected: M): M = apply {
    expected.forEach {
        shouldContain(it.toPair())
    }
    this.forEach {
        expected.shouldContain(it.toPair())
    }
}
```

**Comment by MarkusAmshove** [comment]
This is fixed in 1.58 :)
